### PR TITLE
binary-search-tree: add new exercise

### PR DESCRIPTION
**Pull request by filiptoskovic**

### CODE REVIEW

**Comment by rpottsoh** [comment]
@filiptoskovic thanks for working on this.  I'll take closer look at this and get back to you.  At a quick glance this looks good.

**Comment by filiptoskovic** [comment]
Happy new year! Long time no new exercise, but in December I had literally no spare time.

**Comment by rpottsoh** [comment]
Happy New Year to you too.  I figured you must have been busy.


Sent from my android device.

File: exercises/binary-search-tree/uBinarySearchTreeExample.pas [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+begin
+  if FValue < AVal then
+    DoAdd(AVal, Right)
+  else
+    DoAdd(AVal, Left);
+end;
+
+constructor TBinarySearchTree.Create(AVal: TArray<string>);
+var
+  I, L: Integer;
```
Comment (rpottsoh) [inline exercises/binary-search-tree/uBinarySearchTreeExample.pas]:
```suggestion
  i, L: Integer;
```

File: exercises/binary-search-tree/uBinarySearchTreeExample.pas [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+begin
+  inherited Create;
+  L := length(AVal);
+  if L > 0 then
+  begin
+    FValue := AVal[0];
+    Left := nil;
+    Right := nil;
+
+    for I := 1 to length(AVal) - 1 do
```
Comment (rpottsoh) [inline exercises/binary-search-tree/uBinarySearchTreeExample.pas]:
```suggestion
    for i := 1 to length(AVal) - 1 do
```

File: exercises/binary-search-tree/uBinarySearchTreeTest.pas [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+
+    [Test]
+    [Ignore]
+    procedure greater_number_at_right_node;
+
+    [Test]
+    [Ignore]
+    procedure can_create_complex_tree;
+
+   [Test]
```
Comment (rpottsoh) [inline exercises/binary-search-tree/uBinarySearchTreeTest.pas]:
```suggestion
    [Test]
```

File: exercises/binary-search-tree/uBinarySearchTreeTest.pas [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+    [Ignore]
+    procedure can_sort_single_number;
+
+   [Test]
```
Comment (rpottsoh) [inline exercises/binary-search-tree/uBinarySearchTreeTest.pas]:
```suggestion
    [Test]
```

File: exercises/binary-search-tree/uBinarySearchTreeTest.pas [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+    procedure can_sort_if_second_number_is_same_as_first;
+
+    [Test]
+    [Ignore]
+    procedure can_sort_if_second_number_is_greater_than_first;
+
+    [Test]
+    [Ignore]
+    procedure can_sort_complex_tree;
+
```
Comment (rpottsoh) [inline exercises/binary-search-tree/uBinarySearchTreeTest.pas]:
```suggestion
```

File: exercises/binary-search-tree/uBinarySearchTreeExample.pas [hunk 6]
```diff
@@ -0,0 +1,82 @@
+unit uBinarySearchTree;
+
+interface
+
+type
+  TBinarySearchTree = class
+  private
+    FValue : string;
```
Comment (rpottsoh) [inline exercises/binary-search-tree/uBinarySearchTreeExample.pas]:
```suggestion
    FValue: string;
```

File: exercises/binary-search-tree/uBinarySearchTreeExample.pas [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+    procedure DoAdd(Aval : string; var ABranch : TBinarySearchTree);
+  public
+    Left, Right : TBinarySearchTree;
```
Comment (rpottsoh) [inline exercises/binary-search-tree/uBinarySearchTreeExample.pas]:
```suggestion
    Left, Right: TBinarySearchTree;
```

File: exercises/binary-search-tree/uBinarySearchTreeExample.pas [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+    procedure Add(AVal : string);
```
Comment (rpottsoh) [inline exercises/binary-search-tree/uBinarySearchTreeExample.pas]:
```suggestion
    procedure Add(AVal: string);
```

File: exercises/binary-search-tree/uBinarySearchTreeExample.pas [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+    function SortedData : TArray<string>;
```
Comment (rpottsoh) [inline exercises/binary-search-tree/uBinarySearchTreeExample.pas]:
```suggestion
    function SortedData: TArray<string>;
```

File: exercises/binary-search-tree/uBinarySearchTreeExample.pas [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+    property Data : string read FValue;
```
Comment (rpottsoh) [inline exercises/binary-search-tree/uBinarySearchTreeExample.pas]:
```suggestion
    property Data: string read FValue;
```

File: exercises/binary-search-tree/uBinarySearchTreeExample.pas [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+    constructor Create(AVal : TArray<string>);
```
Comment (rpottsoh) [inline exercises/binary-search-tree/uBinarySearchTreeExample.pas]:
```suggestion
    constructor Create(AVal: TArray<string>);
```

File: exercises/binary-search-tree/uBinarySearchTreeExample.pas [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+begin
+  if Assigned(ABranch) then
+    ABranch.Add(AVal)
+  else
+    ABranch := TBinarySearchTree.Create([AVal]);
+end;
+
+function TBinarySearchTree.SortedData: TArray<string>;
+var
+  List : TList<string>;
```
Comment (rpottsoh) [inline exercises/binary-search-tree/uBinarySearchTreeExample.pas]:
```suggestion
  List: TList<string>;
```

**Comment by rpottsoh** [comment]
This is looking good.  Let me know when you are done and I will merge.

**Pull request merged**